src/Utilities/Compiler/Extensions/DiagnosticExtensions.cs

```diff
@@ -184,15 +188,22 @@ internal static class DiagnosticExtensions
 
             DiagnosticSeverity? GetEffectiveSeverity()
             {
+#if !CODEANALYSIS_V3_OR_BETTER
                 if (s_syntaxTreeDiagnosticOptionsProperty == null)
                 {
                     return rule.DefaultSeverity;
                 }
+#endif
 
                 ReportDiagnostic? overriddenSeverity = null;
                 foreach (var tree in compilation.SyntaxTrees)
                 {
-                    var options = (ImmutableDictionary<string, ReportDiagnostic>)s_syntaxTreeDiagnosticOptionsProperty.GetValue(tree)!;
+                    var options =
+#if CODEANALYSIS_V3_OR_BETTER
+                        tree.DiagnosticOptions;
```

Comment:
SyntaxTree.DiagnosticOptions has actually been deprecated in future Roslyn versions and always returns empty options. I need to fix up the logic here.

Reply:
https://github.com/dotnet/roslyn-analyzers/issues/4224